nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/PackageManagerImpl.java

```diff
@@ -1096,24 +1103,48 @@ public void checkOptionalDependenciesOnInstalledPackages(DependencyResolution re
+                    }
+                    if (!hasAnInstalledMatch) {
+                        // if optional dependency is not already installed, check if it is currently being installed
                         for (String pkgId : res.getOrderedPackageIdsToInstall()) {
                             DownloadablePackage pkgToInstall = getPackage(pkgId);
                             if (matchDependency(pkgOptDep, pkgToInstall)) {
-                                log.info(String.format(
+                                res.logInfo(String.format(
                                         "As package '%s' has an optional dependency on package '%s' currently being installed, it will be reinstalled.",
                                         installedPkg, pkgToInstall));
                                 packagesToReinstall.add(installedPkg);
                             }
                         }
                     } else {
-                        for (String pkgId : res.getOrderedPackageIdsToRemove()) {
-                            DownloadablePackage pkgToRemove = getPackage(pkgId);
-                            if (matchDependency(pkgOptDep, pkgToRemove)) {
-                                log.info(String.format(
-                                        "As package '%s' has an optional dependency on package '%s' currently being uninstalled, it will be reinstalled.",
-                                        installedPkg, pkgToRemove));
-                                packagesToReinstall.add(installedPkg);
+                        // find matching currently installing version (upgrade)
+                        boolean hasAnUpgradingMatch = false;
+                        for (String pkgId : res.getOrderedPackageIdsToInstall()) {
+                            DownloadablePackage pkgToInstall = getPackage(pkgId);
+                            if (pkgToInstall.getName().equals(pkgOptDep.getName())
+                                    && pkgOptDep.getVersionRange().matchVersion(pkgToInstall.getVersion())) {
+                                hasAnUpgradingMatch = true;
+                                break;
+                            }
+                        }
+                        if (!hasAnUpgradingMatch) {
+                            // if optional dependency is already installed and is not currently being upgraded, check if
+                            // it is currently being uninstalled
+                            for (String pkgId : res.getOrderedPackageIdsToRemove()) {
+                                DownloadablePackage pkgToRemove = getPackage(pkgId);
+                                if (matchDependency(pkgOptDep, pkgToRemove)) {
+                                    res.logInfo(String.format(
+                                            "As package '%s' has an optional dependency on package '%s' currently being uninstalled, it will be reinstalled.",
+                                            installedPkg, pkgToRemove));
```

Comment:
Rather delay the end-user interpretation / log output and keep it as the responsibility of toString() method:

                                   res.addUninstallOptDept(installedPkg, pkgToRemove);
                                    if (log.isDebugEnabled()) {
                                        log.debug(new Throwable(),...);
                                    }
                                    packagesToReinstall.add(installedPkg);
                                    break;